keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
 The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of the container runtime that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
 
 ### Services
+Services depend on `--services-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. A dualstack cluster can have a single cidr in this flag. That means the cluster will be to egress traffic on both ip families, but will *not* be ingress traffic on both ip families.
+4. When upgrading a cluster to dualstack, the first element of the `--services-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
```

Comment:
just curiosity only, this is about implementation so unrelated to the KEP, but how do we detect this? validating against the the well known `kubernetes` default service IP family?
do we validate and fail in all the components that have the `--services-cluster-ip-range` flag?

Reply:
We don't validate against this, as you said there is no way we can do that. We will have to document it clearly.

Reply:
We can make apiserver validate this; if it starts up and finds that `kubernetes.default` already exists but is of the wrong IP family, then it knows that you broke the rule, and it could refuse to start. You can't even easily thwart that check by deleting the old `kubernetes.default` before restarting, because apiserver will immediately recreate it if you try. You'd have to stop apiserver and then delete it with `etcdctl` or something. And at that point you deserve to lose.

Reply:
There are still some folks who want to change kubernetes.default to support ExternalName, and I am somewhat sympathetic to the use-case.  In that case, we don't have much else to stand on.  If we had all the time in the world we could devise a way for apiserver to snapshot some state into etcd (like the port allocator) and check it on startup.  Is that worth it?  Do we have anything ELSE we would want to do with that?